### ISSUE TITLE
DeepSeek: fix reduced-row device sampling regressions

### ISSUE THREAD

**Issue by yieldthought** [tenstorrent/tt-metal#41357]
## Summary
Reduced-row DeepSeek device sampling regressed in two places while plumbing `max_users_per_row` through the demo/runtime:

1. The traced path still requires a 32-slot sampling batch and now throws for `max_users_per_row < 32`.
2. The non-traced path pads logits to the sampler's 32-slot row width, but token extraction still unwraps the result as if the logical batch stayed unpadded.

## Impact
This breaks or makes unsafe the new reduced-row (`max_users_per_row=8`) demo/runtime path that was added to exercise batch-8 decode.

### 1. Trace path hard-fails for reduced-row sampling
`models/demos/deepseek_v3/tt/generator.py` now builds sampling args with `max_batch_size=self.batch_size_per_row`, but `models/common/sampling/tt_sampling.py` still rounds that to a 32-slot device batch. When tracing is enabled, `_sample_tokens_device()` rejects an 8-row logits tensor instead of padding it.

Relevant paths:
- `models/demos/deepseek_v3/tt/generator.py:768`
- `models/common/sampling/tt_sampling.py:123`
- `models/demos/deepseek_v3/demo/test_demo.py:117`

### 2. Reduced-row non-traced sampling no longer has a safe unwrap path
The non-traced path pads logits up to the sampler's 32-slot row width before sampling, but `_tokens_from_device()` still tries to decode the result using the unpadded `batch_size_per_row`. If the sampled output keeps the padded logical batch, the layout-specific branches no longer match and the code falls back to flattening the raw composed tensor and slicing the first `N` values. That is not a safe way to recover the active-user tokens from the mesh-composed output.

Relevant paths:
- `models/demos/deepseek_v3/tt/generator.py:806`
- `models/demos/deepseek_v3/tt/generator.py:812`
- `models/demos/deepseek_v3/tt/generator.py:814`

## Repro
One concrete repro for the first problem is running the DeepSeek demo with reduced rows and tracing enabled, for example:

```bash
python models/demos/deepseek_v3/demo/demo.py \
  --max-users-per-row 8 \
  --enable-trace \
  --sample-on-device
```

## Expected
Reduced-row device sampling should behave the same as the 32-user path, including with tracing enabled, and token extraction should explicitly unwrap/preserve only the active users in the correct order.
